src/drivers/intel/ssp/ssp.c

```diff
@@ -794,19 +803,76 @@ static int ssp_get_hw_params(struct dai *dai,
 	return 0;
 }
 
-/* SSP hw params */
+/*
+ * enabled clock MCLK/BCLK to support codec that need MCLK/BCLK
+ * This is called before trigger(COMP_TRIGGER_START)
+ */
 static int ssp_hw_params(struct dai *dai,
 			 struct sof_ipc_stream_params  *params)
 {
-	dai_info(dai, "ssp_hw_params()");
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+	int ret;
 
+	if (ssp->state[SOF_IPC_STREAM_CAPTURE] < COMP_STATE_PREPARE &&
+	    ssp->state[SOF_IPC_STREAM_PLAYBACK] < COMP_STATE_PREPARE) {
+		dai_err(dai, "ssp_hw_params(): Please set config first");
+		return -EINVAL;
```

Comment:
While re-checking the code, that test seems really odd to me. Usually the hw_params is done *BEFORE* the prepare step. here it's inverted. That's inconsistent and points to a really odd state machine for the dai.

Reply:
> While re-checking the code, that test seems really odd to me. Usually the hw_params is done _BEFORE_ the prepare step. here it's inverted. That's inconsistent and points to a really odd state machine for the dai.

`ssp->state = COMP_STATE_PREPARE` is set in `ssp_set_config`. We didn't handle COMP_TRIGGER_PREPARE in `ssp_trigger`. That says ssp->state doesn't match PCM state.